fuzz/ml-kem.c

```diff
@@ -0,0 +1,685 @@
+{
+    uint16_t keysize;
+    uint16_t modulus = 6;
+
+    *buf = consume_uint16t(*buf, len, &keysize);
+    if (*buf == NULL)
+        return 0;
+
+    /*
+     * select from sizes
+     * ML-KEM-512, ML-KEM-768, and ML-KEM-1024
+     * also select some invalid sizes to trigger
+     * error paths
+     */
+    if (only_valid)
+        modulus = 3;
+
+    /*
+     * Note, keylens for valid values (cases 0-2)
+     * are taken based on input values from our unit tests
+     */
+    switch (keysize % modulus) {
+    case 0:
+        *keytype = "ML-KEM-512";
+        *keylen = 800;
+        break;
+    case 1:
+        *keytype = "ML-KEM-768";
+        *keylen = 1172;
+        break;
+    case 2:
+        *keytype = "ML-KEM-1024";
+        *keylen = 1568;
+        break;
+    case 3:
+        /* select invalid alg */
+        *keytype = "ML-KEM-13";
+        *keylen = 13;
+        break;
+    case 4:
+        /* Select valid alg, but bogus size */
+        *keytype = "ML-KEM-1024";
+        *buf = consume_uint16t(*buf, len, &keysize);
+        *keylen = (size_t)keysize;
+        *keylen %= 2048; /* size to our key buffer */
+        break;
+    default:
+        *keytype = NULL;
+        *keylen = 0;
+        break;
+    }
+    return 1;
+}
+
+/**
+ * @brief Creates an ML-KEM raw key from a buffer.
+ *
+ * This function selects a key type and size from the buffer, generates
+ * a random key of the appropriate length, and creates either a public
+ * or private ML-KEM key using OpenSSL's EVP_PKEY interface.
+ *
+ * @param buf   Pointer to the buffer pointer; updated after reading.
+ * @param len   Pointer to the remaining buffer size; updated accordingly.
+ * @param key1  Pointer to store the generated EVP_PKEY key (public or private).
+ * @param key2  Unused parameter (reserved for future use).
+ *
+ * @note The generated key is allocated using OpenSSL's EVP_PKEY functions
+ *       and should be freed appropriately using `EVP_PKEY_free()`.
+ */
+static void create_mlkem_raw_key(uint8_t **buf, size_t *len,
+                                 void **key1, void **key2)
+{
+    EVP_PKEY *pubkey;
+    char *keytype = NULL;
+    size_t keylen = 0;
+    uint8_t key[2048];
+    int pub = 0;
+
+    if (!select_keytype_and_size(buf, len, &keytype, &keylen, 0))
+        return;
+
+    /*
+     * Select public or private key creation based on the low order
+     * bit of the next buffer value
+     */
+    if ((*buf)[0] & 0x1)
+        pub = 1;
+
+    /*
+     * libfuzzer provides by default up to 4096 bit input
+     * buffers, but its typically much less (between 1 and 100 bytes)
+     * so use RAND_bytes here instead
+     */
+    if (!RAND_bytes(key, keylen))
+        return;
+
+    /*
+     * Try to generate either a raw public or private key using random data
+     * Because the input is completely random, its likely this operation will
+     * fail, but it will still exercise the code paths below, which is what we
+     * want the fuzzer to do
```

Comment:
It will fail with extremely high probability (never succeed in lifetime of universe), because valid keys are highly constrained. To get valid keys, you need to actually use the ML-KEM genkey API, and then perhaps mutate them a bit, but you will never accidentally stumble onto one.

Reply:
right, and thats intentional as it exercises the error paths in the mlkem provider.  There is a separate function keygen_mlkem_real_key that generates keys using the keygen api, which is meant to produce actual working keys.

Reply:
OK. No changes required, but that "likely" could be changed to "effectively certain".